sensor/common/sensor/buffered_stream.go

```diff
@@ -31,18 +38,30 @@ func (s bufferedStream) Send(msg *central.MsgFromSensor) error {
 	return nil
 }
 
-func NewBufferedStream(stream messagestream.SensorMessageStream, msgC chan *central.MsgFromSensor, stopC concurrency.ReadOnlyErrorSignal) (messagestream.SensorMessageStream, <-chan error) {
+func NewBufferedStream(stream messagestream.SensorMessageStream, msgC chan *central.MsgFromSensor, stopC concurrency.ReadOnlyErrorSignal) (messagestream.SensorMessageStream, <-chan error, func() error) {
 	// if the capacity of the buffer is zero then we just return the inner stream
 	if cap(msgC) == 0 {
-		return stream, nil
+		return stream, nil, nil
 	}
 	ret := bufferedStream{
 		buffer: msgC,
 		stopC:  stopC,
 		stream: stream,
 	}
 	errC := ret.run()
-	return ret, errC
+	return ret, errC, func() error {
+		for {
+			select {
+			case _, ok := <-errC:
+				if !ok {
+					return nil
+				}
+			case <-time.After(stopTimeout):
+				// If we reach this timeout we could have a deadlock in the gRPC stream
+				return errors.New("timeout waiting for the buffered stream to stop")
+			}
+		}
+	}
```

Comment:
This looks reasonable but a bit weird to me... You return `errC`, so the func that is being returned as a third param can be written independently. 
However, I suspect that your intention was to force the caller to check for the timeout (because it would be easy to skip that). If that is the case, then maybe we should prevent the caller access to `errC` but instead change the third func parameter so that is has to be used? I don't have a cristal-clear idea yet, but here is something I was thinking about:

```suggestion
	return ret, func(foreachErr func(e error)) error {
		for {
			select {
			case err, ok := <-errC:
				if !ok {
					return nil
				}
				foreachErr(err)
			case <-time.After(stopTimeout):
				// If we reach this timeout we could have a deadlock in the gRPC stream
				return errors.New("timeout waiting for the buffered stream to stop")
			}
		}
	}
```

Reply:
^^ this is non-blocking comment, so if you planned to merge this PR as it is, I won't be objecting.

Reply:
🤔 The objective of this third param is to have a way to consume any leftover messages in the channel and wait for errC to be closed before closing the gRPC stream. Otherwise we will have a race between `Send` and `CloseSend`. The tricky part is that the first error received in `errC` (second param) will trigger the stopping logic. Basically:
```mermaid
sequenceDiagram
    participant conn as CentralConnection
    participant sender as Sender
    participant buff as BufferedStream
    participant grpc as gRPCStream
    grpc->>buff: some error 
    buff->>sender: read from errC 
    activate sender
    sender->>buff: close errC
    activate buff
    sender->>sender: stop
    sender->>conn: stop
    deactivate sender
    conn->>grpc: CloseSend
    Note over buff, grpc: Since we could have messages buffered,<br/> the BufferedStream can call Send again after CloseSend is called
    buff->>grpc: Send
    deactivate buff
```

With these changes:
```mermaid
sequenceDiagram
    participant conn as CentralConnection
    participant sender as Sender
    participant buff as BufferedStream
    participant grpc as gRPCStream
    grpc->>buff: some error 
    buff->>sender: read from errC 
    activate sender
    sender->>buff: close errC
    activate buff
    sender->>sender: stop
    buff->>sender: wait for errC to be closed
    deactivate buff
    sender->>conn: stop
    deactivate sender
    conn->>grpc: CloseSend
```
Initially I had a function in the `centralSender` like:
```golang
for {
	select {
	case _, ok := <-errC:
		if !ok {
			return nil
		}
	case <-time.After(stopTimeout):
		// If we reach this timeout we could have a deadlock in the gRPC stream
		return errors.New("timeout waiting for the buffered stream to stop")
	}
}
``` 
I thought it was more readable to return that function as a third parameter but technically the responsibility of exhausting the `errC` and wait until it's closed should be on the `centralSender`.

Reply:
We could also make it work with your proposal but somehow we would need to trigger the stop process from the `foreachErr` function on the first error and have that running from the begging in a goroutine that then we will have to sync. I feel it might be more complicated to make it work.

Reply:
> The objective of this third param is to have a way to consume any leftover messages in the channel and wait for errC to be closed before closing the gRPC stream.

Let's capture that in a comment in the code and this is ✅ for me 👍

Reply:
Yeah, let’s do that. I was trying to figure out how to do it without too much refactoring and in all implementations I had this feeling that we need a comment to explain this.